File: ios/Classes/Core/Handler/BaseMethodHandler.swift
import Flutter

protocol MethodHandler  {
    associatedtype MethodT
    func onMethod(method: MethodT, result: @escaping FlutterResult) throws
}

// MARK: - BaseMethodHandler
class BaseMethodHandler<T>: NSObject, FlutterPlugin, MethodHandler {
    typealias MethodT = T
    
    let parser: MethodParser<T>
    
    init(parser: MethodParser<T>) {
        self.parser = parser
    }
    
    @objc static func register(with registrar: FlutterPluginRegistrar) {
        // Refer to `SwiftFlutterKlarnaInappSdk`
    }
        
    final func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        do {
            let method = try parser.parse(call: call)
            if let m = method {
                try onMethod(method: m, result: result)
            } else {
                result(FlutterMethodNotImplemented)
            }
        } catch let error {
            result(FlutterError.init(code: ResultError.pluginMethodError.rawValue, message: call.method, details: error))
        }
    }
    
    func onMethod(method: T, result: @escaping FlutterResult) throws {
        
    }
    
}

File: packages/klarna_mobile_sdk_flutter/android/src/main/kotlin/com/klarna/mobile/sdk/flutter/core/handler/BaseMethodHandler.kt
package com.klarna.mobile.sdk.flutter.core.handler

import com.klarna.mobile.sdk.flutter.ResultError
import com.klarna.mobile.sdk.flutter.core.method.MethodParser
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal abstract class BaseMethodHandler<T>(val parser: MethodParser<T>) : MethodChannel.MethodCallHandler {
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            val method = parser.parse(call)
            if (method != null) {
                onMethod(method, result)
            } else {
                result.notImplemented()
            }
        } catch (e: Throwable) {
            result.error(ResultError.PLUGIN_METHOD_ERROR.errorCode, call.method, e.message)
        }
    }

    abstract fun onMethod(method: T, result: MethodChannel.Result)
}